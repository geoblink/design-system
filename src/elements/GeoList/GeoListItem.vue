<template>
  <component
    :is="wrapperTag"
    v-bind="$attrs"
    :class="[{
      'geo-list-item': true,
      'geo-list-item--disabled': disabled
    }]"
    v-on="computedListeners"
  >
    <div class="geo-list-item__label-and-accessory-container">
      <div class="geo-list-item__icon-and-label">
        <div
          v-if="icon"
          class="geo-list-item__icon-and-label__icon-container"
        >
          <font-awesome-icon
            :icon="icon"
            class="geo-list-item__icon-and-label__icon-container__icon"
            aria-hidden
            fixed-width
          />
        </div>
        <div class="geo-list-item__icon-and-label__label">
          <!-- @slot Use this slot to customize rows's main content -->
          <slot />
        </div>
      </div>
      <div
        v-if="isTrailingItemWrapperVisible"
        class="geo-list-item__trailing-accessory-items"
      >
        <!-- @slot Use this slot to add more items to the trailing edge of this row -->
        <slot name="trailingAccessoryItem" />

        <template v-if="disabled && disabledIcon">
          <font-awesome-icon
            :icon="disabledIcon"
            class="geo-list-item__icon-and-label__icon-container__icon"
            aria-hidden
            fixed-width
          />
        </template>
      </div>
    </div>
    <div
      v-if="$slots.description"
      :class="{
        'geo-list-item__description': true,
        'geo-list-item__description--spaced': !!icon
      }"
    >
      <!-- @slot Use this slot to add a description for the item -->
      <slot name="description" />
    </div>
    <geo-tooltip
      v-if="disabledTooltipContent"
      :position="disabledTooltipPosition"
    >
      {{ disabledTooltipContent }}
    </geo-tooltip>
  </component>
</template>

<script>
import extendedListenersMixin from '@/mixins/extendedListenersMixin'
import { enumPropertyFactory } from '../../utils/enumPropertyFactory'
import { POSITIONS } from '../GeoTooltip/GeoTooltip.constants'

/**
 * `GeoListItem` is a component designed to build vertical lists which fit
 * properly in a `GeoBorderedBox`.
 *
 * It can be customized in several ways, adding icons to the leading edge of the
 * row or more complex elements (like form inputs) to the trailing edge.
 */
export default {
  name: 'GeoListItem',
  status: 'ready',
  release: '4.0.0',
  mixins: [extendedListenersMixin],
  props: {
    /**
     * Optional Font Awesome 5 icon to be displayed next to the entry's label,
     * on the left.
     *
     * See [vue-fontawesome](https://www.npmjs.com/package/@fortawesome/vue-fontawesome#explicit-prefix-note-the-vue-bind-shorthand-because-this-uses-an-array)
     * for more info about this.
     */
    icon: {
      type: Array,
      required: false
    },

    /**
     * Whether this item is disabled or not. When disabled it will be displayed
     * greyed out.
     *
     * **Note:** Listeners won't be affected by this property.
     */
    disabled: {
      type: Boolean,
      default: false
    },

    /**
     * Optional Font Awesome 5 icon to be displayed next to the entry's label,
     * on the right when the item is disabled.
     *
     * See [vue-fontawesome](https://www.npmjs.com/package/@fortawesome/vue-fontawesome#explicit-prefix-note-the-vue-bind-shorthand-because-this-uses-an-array)
     * for more info about this.
     */
    disabledIcon: {
      type: Array,
      required: false
    },

    /**
     * Text to be displayed inside a tooltip when the item is disabled.
     */
    disabledTooltipContent: {
      type: String,
      required: false
    },

    /**
     * Position for the tooltip with the disabled tooltip content
     */
    disabledTooltipPosition: enumPropertyFactory({
      componentName: 'GeoListItem',
      propertyName: 'disabledTooltipPosition',
      enumDictionary: POSITIONS,
      defaultValue: POSITIONS.trailing
    }),

    /**
     * HTML tag that will be used as a wrapper of the GeoListItem
     */
    wrapperTag: {
      type: String,
      default: 'div'
    }
  },
  computed: {
    isTrailingItemWrapperVisible () {
      return this.$slots.trailingAccessoryItem || (this.disabled && this.disabledIcon)
    }
  }
}

</script>
