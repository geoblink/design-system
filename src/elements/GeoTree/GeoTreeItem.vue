<template>
  <li
    class="geo-tree-item"
  >
    <geo-list-item
      :class="{
        'geo-tree-item--rotated-icon': hasToRotateIcon,
        'geo-tree-item--single': isSingleItem
      }"
      :icon="categoryIcon"
      @click="handleClick"
    >
      <label class="geo-tree-item__label">
        <geo-highlighted-string
          :highlighted-chars="category.matches"
          :reference-string="category[keyForLabel]"
        />
        <span
          v-if="!isSingleItem"
          class="geo-tree-item__total-items"
        >
          ({{ totalCategoryChildren }})
        </span>
        <span
          v-if="category.description"
          class="geo-tree-item__description"
        >
          <font-awesome-icon
            :icon="descriptionIcon"
            fixed-width
          />
          <geo-tooltip>
            {{ category.description }}
          </geo-tooltip>
        </span>
      </label>
      <template slot="trailingAccessoryItem">
        <span>
          <slot
            name="trailingAccessoryAction"
            :item="category"
          />
        </span>
        <input
          :id="category[keyForId]"
          :checked="isChecked"
          :indeterminate.prop="isIndeterminate"
          :disabled="isInputDisabled"
          type="checkbox"
          @click.stop
          @input="handleCheck(category, $event.target.checked)"
        >
      </template>
    </geo-list-item>
    <ul
      v-if="isExpanded"
      class="geo-tree-item__list"
    >
      <draggable
        :list="category[keyForSubcategory]"
        :group="draggableGroup"
        :sort="false"
        :disabled="!draggableGroup"
        :filter="`.${dragClassToIgnore}`"
        drag-class="geo-tree__dragger"
        ghost-class="geo-tree__ghost"
        @start="startDrag($event)"
        @end="endDrag($event)"
        @change="changeDrag($event, category)"
      >
        <geo-tree-item
          v-for="subcategory in category[keyForSubcategory]"
          :key="subcategory[keyForId]"
          :class="dragClassToIgnore"
          :data-test="`subcategory-${subcategory[keyForId]}`"
          :category="subcategory"
          :key-for-id="keyForId"
          :key-for-label="keyForLabel"
          :key-for-subcategory="keyForSubcategory"
          :expanded-categories="expandedCategories"
          :checked-items="checkedItems"
          :collapsed-icon="collapsedIcon"
          :expanded-icon="expandedIcon"
          :description-icon="descriptionIcon"
          :draggable-group="draggableGroup"
          @check-item="handleCheckChildItem"
          @check-folder="handleCheckChildFolder"
          @click="handleClick"
          @toggleExpand="toggleExpand"
          @start-drag="startDrag($event)"
          @end-drag="endDrag($event)"
          @change-drag="emitChangeDrag($event)"
        >
          <template
            slot="trailingAccessoryAction"
            slot-scope="{ item }"
          >
            <slot
              name="trailingAccessoryAction"
              :item="item"
            />
          </template>
        </geo-tree-item>
      </draggable>
    </ul>
  </li>
</template>

<script>
import _ from 'lodash'
import Draggable from 'vuedraggable'

import GeoTreeMixin from './GeoTreeMixin'

export default {
  name: 'GeoTreeItem',
  status: 'ready',
  release: '29.9.0',
  components: {
    Draggable
  },
  mixins: [GeoTreeMixin],
  props: {
    category: {
      type: Object,
      required: true
    },
    /**
     * Key to access to the identifier of the item
     */
    keyForId: {
      type: String,
      required: false,
      default: 'id'
    },
    /**
     * Key to access to the label of the item
     */
    keyForLabel: {
      type: String,
      required: false,
      default: 'label'
    },
    /**
     * Key to access to the children items of the category
     */
    keyForSubcategory: {
      type: String,
      required: false,
      default: 'subcategory'
    },
    /**
     * List of all the items checked, it's an object with truthy keys
     */
    checkedItems: {
      type: Object,
      required: true
    },
    /**
     * Icon used to alert about some extra info displayed in a popover
     */
    descriptionIcon: {
      type: Array,
      default: () => []
    },
    /*
    * List of categories that should appear expanded
    * */
    expandedCategories: {
      type: Object,
      required: true
    },
    /**
     * Optional Font Awesome 5 icon to use as collapsed icon
     *
     * See [vue-fontawesome](https://www.npmjs.com/package/@fortawesome/vue-fontawesome#explicit-prefix-note-the-vue-bind-shorthand-because-this-uses-an-array)
     * for more info about this.
     */
    collapsedIcon: {
      type: Array,
      required: true,
      default: function () {
        return ['fal', 'chevron-right']
      }
    },
    /**
     * Optional Font Awesome 5 icon to use as expanded icon
     *
     * See [vue-fontawesome](https://www.npmjs.com/package/@fortawesome/vue-fontawesome#explicit-prefix-note-the-vue-bind-shorthand-because-this-uses-an-array)
     * for more info about this.
     */
    expandedIcon: {
      type: Array,
      required: false
    }
  },
  data () {
    return {
      isShowActionButton: false
    }
  },
  computed: {
    categoryIcon () {
      if (this.isSingleItem) return null

      return this.isExpanded
        ? this.expandedIcon || this.collapsedIcon
        : this.collapsedIcon
    },
    isIndeterminate () {
      const isSomeChildSelected = category => {
        return _.some(category[this.keyForSubcategory], subCategory => {
          if (subCategory[this.keyForSubcategory]) return isSomeChildSelected(subCategory)

          return !!this.checkedItems[subCategory[this.keyForId]]
        })
      }

      return !this.isChecked && isSomeChildSelected(this.category)
    },
    isChecked () {
      if (this.isInputDisabled) return false

      const allAreChildrenSelected = category => {
        const selectableChildren = this.getSelectableChildrenForCategory(category)
        return _.every(selectableChildren, subCategory => {
          if (_.size(subCategory[this.keyForSubcategory])) return allAreChildrenSelected(subCategory)

          return !!this.checkedItems[subCategory[this.keyForId]]
        })
      }
      return !this.isSingleItem
        ? allAreChildrenSelected(this.category)
        : !!this.checkedItems[this.category[this.keyForId]]
    },
    totalCategoryChildren () {
      const sumOfChildren = category => _.size(category[this.keyForSubcategory]) + _.sumBy(category[this.keyForSubcategory], sumOfChildren)

      return sumOfChildren(this.category)
    },
    totalSelectableCategoryChildren () {
      const sumOfSelectableChildren = category => {
        const selectableChildren = this.getSelectableChildrenForCategory(category)
        return _.size(selectableChildren) + _.sumBy(category[this.keyForSubcategory], sumOfSelectableChildren)
      }

      return sumOfSelectableChildren(this.category)
    },
    isExpanded () {
      return !!this.expandedCategories[this.category[this.keyForId]]
    },
    isSingleItem () {
      return _.isNil(this.category[this.keyForSubcategory])
    },
    hasToRotateIcon () {
      return this.isExpanded && !this.expandedIcon
    },
    isInputDisabled () {
      return !this.isSingleItem && this.totalSelectableCategoryChildren === 0
    }
  },
  methods: {
    /**
     * On list item click
     */
    handleClick () {
      if (this.isSingleItem) {
        this.handleCheck(this.category, !this.isChecked)
      } else {
        this.toggleExpand(this.category)
      }
    },
    handleCheckChildItem (category, isChecked, isDelegated) {
      this.$emit('check-item', category, isChecked, isDelegated)
    },
    handleCheckChildFolder (category, isChecked, isDelegated) {
      this.$emit('check-folder', category, isChecked, isDelegated)
    },
    /**
     * To check all items of a category
     */
    handleCheck (category, isChecked, isDelegated = false) {
      if (category[this.keyForSubcategory]) {
        if (_.size(category[this.keyForSubcategory])) { // We don't want to emit event for empty folders
          this.$emit('check-folder', _.omit(category, this.keyForSubcategory), isChecked, isDelegated)
        }
        _.forEach(category[this.keyForSubcategory], (innerCategory) => {
          this.handleCheck(innerCategory, isChecked, true)
        })
      } else {
        this.$emit('check-item', category, isChecked, isDelegated)
      }
    },
    toggleExpand (category) {
      this.$emit('toggleExpand', category)
    },
    getSelectableChildrenForCategory (category) {
      const keyForSubcategory = this.keyForSubcategory
      return getSelectableChildrenRecursively(category)

      function getSelectableChildrenRecursively (currentCategory) {
        return _.reject(currentCategory[keyForSubcategory], subCategory => {
          return subCategory[keyForSubcategory] &&
            (!_.size(subCategory[keyForSubcategory]) || !_.size(getSelectableChildrenRecursively(subCategory)))
        })
      }
    }
  }
}
</script>
