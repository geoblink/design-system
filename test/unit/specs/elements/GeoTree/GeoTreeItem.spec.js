import _ from 'lodash'
import Vue from 'vue'
import { shallowMount } from '@vue/test-utils'
import GeoTreeItem from '@/elements/GeoTree/GeoTreeItem.vue'
import GeoListItem from '@/elements/GeoList/GeoListItem.vue'

const CATEGORY = {
  id: 'fruits',
  label: 'Fruits',
  subcategories: [
    {
      id: 'tropical-fruits',
      label: 'Tropical fruits',
      subcategories: [
        {
          id: 'pineapple',
          label: 'Pineapple'
        },
        {
          id: 'banana',
          label: 'Banana'
        },
        {
          id: 'coconut',
          label: 'Coconut'
        },
        {
          id: 'avocado',
          label: 'Avocado'
        }
      ]
    }
  ]
}

const EXPANDED_CATEGORY = {
  id: 'fruits',
  label: 'Fruits',
  isExpanded: true,
  subcategories: [
    {
      id: 'tropical-fruits',
      label: 'Tropical fruits',
      isExpanded: true,
      subcategories: [
        {
          id: 'pineapple',
          label: 'Pineapple'
        },
        {
          id: 'banana',
          label: 'Banana'
        },
        {
          id: 'coconut',
          label: 'Coconut'
        },
        {
          id: 'avocado',
          label: 'Avocado'
        }
      ]
    }
  ]
}

describe.only('GeoTreeItem', () => {
  const getWrapper = (props = {}) => shallowMount(GeoTreeItem, {
    propsData: _.assign(
      {},
      {
        keyForLabel: 'label',
        keyForSubcategory: 'subcategories',
        keyForId: 'id',
        category: props.category || CATEGORY,
        checkedItems: {}
      },
      props
    ),
    stubs: {
      'geo-list-item': GeoListItem,
      'font-awesome-icon': true,
      'geo-highlighted-string': true,
      'geo-tooltip': true,
      'geo-tree-item': GeoTreeItem
    }
  })

  it('should render the total subcategories number if it has subcategories', () => {
    const wrapper = getWrapper()

    expect(wrapper.find('.geo-tree-item__total-items').exists()).toBe(true)
    expect(wrapper.find('.geo-tree-item__total-items').text()).toBe(`(${5})`)
  })

  it('should render a list of subcategories on category click (it it has subcategories)', async () => {
    const wrapper = getWrapper()

    wrapper.find('.geo-tree-item').trigger('click')
    await Vue.nextTick()
    await wrapper.vm.$forceUpdate()

    expect(wrapper.vm.category.isExpanded).toBe(true)

    wrapper.find('[data-test="subcategory-tropical-fruits"]').trigger('click')
    await Vue.nextTick()

    expect(_.find(wrapper.vm.category.subcategories, { id: 'tropical-fruits' }).isExpanded).toBe(true)
  })

  it('should render a icon next to the category if the category has a description', () => {
    CATEGORY.description = 'Description to show'
    const wrapper = getWrapper({ category: CATEGORY })

    expect(wrapper.find('.geo-tree-item__description').exists()).toBe(true)
    expect(wrapper.find('.geo-tree-item__description').text()).toBe(CATEGORY.description)
  })

  it('should emit the checked item when a category without children nodes is checked', async () => {
    const wrapper = getWrapper({
      category: EXPANDED_CATEGORY
    })

    const itemToCheck = wrapper.find('#avocado')
    itemToCheck.setChecked()
    await itemToCheck.trigger('input')

    expect(wrapper.emitted().check.length).toBe(1)
    expect(wrapper.emitted().check[0]).toEqual([{ id: 'avocado', label: 'Avocado' }, true])
  })

  it('should emit all children nodes when a category with children nodes is checked', async () => {
    const wrapper = getWrapper({
      category: EXPANDED_CATEGORY
    })

    const itemToCheck = wrapper.find('#tropical-fruits')
    itemToCheck.setChecked()
    await itemToCheck.trigger('input')

    expect(wrapper.emitted().check.length).toBe(4)

    const subcategoriesToCheck = _.find(EXPANDED_CATEGORY.subcategories, { id: 'tropical-fruits' })

    _.forEach(subcategoriesToCheck.subcategories, (subcategory, index) => {
      expect(wrapper.emitted().check[index]).toEqual([{ id: subcategory.id, label: subcategory.label }, true])
    })
  })

  it('should mark as check all items passed in checkedItems prop', async () => {
    const wrapper = getWrapper({
      category: EXPANDED_CATEGORY,
      checkedItems: { pineapple: true, coconut: true }
    })

    expect(wrapper.find('#pineapple').element.checked).toBe(true)
    expect(wrapper.find('#coconut').element.checked).toBe(true)
  })

  it.only('should emit as uncheck pre checked items on checking', async () => {
    const wrapper = getWrapper({
      category: EXPANDED_CATEGORY,
      checkedItems: { pineapple: true, coconut: true }
    })

    const itemToCheck1 = wrapper.find('#pineapple')
    itemToCheck1.setChecked(false)
    await itemToCheck1.trigger('input')

    const itemToCheck2 = wrapper.find('#coconut')
    itemToCheck2.setChecked(false)
    await itemToCheck2.trigger('input')

    expect(wrapper.emitted().check.length).toBe(2)
    expect(wrapper.emitted().check[0]).toEqual([{ id: 'pineapple', label: 'Pineapple' }, false])
    expect(wrapper.emitted().check[1]).toEqual([{ id: 'coconut', label: 'Coconut' }, false])
  })
})
