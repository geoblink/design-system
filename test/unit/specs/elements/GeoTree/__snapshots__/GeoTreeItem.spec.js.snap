// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GeoTreeItem should render correctly 1`] = `
<li class="geo-tree-item">
  <geo-list-item-stub icon="fal,chevron-right"><label>
      <geo-highlighted-string-stub reference-string="Fruits"></geo-highlighted-string-stub> <span class="geo-tree-item__total-items">
        (5)
      </span>
      <!---->
    </label>
    <!----> <template></template></geo-list-item-stub>
  <!---->
</li>
`;
